ICE: `unreachable` in compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
struct C;

struct S<T>(Option<T>);

trait Tr {
    type Out;
}

impl<T> Clone for S<C>
where
    S<T>: Tr<Out = T>,
{
    fn clone(&self) -> Self {
        *self
    }
}

fn main() {}
```

A mutant of [`tests/ui/associated-types/issue-25700-1.rs`](https://github.com/rust-lang/rust/blob/c90bcb9571b7aab0d8beaa2ce8a998ffaf079d38/tests/ui/associated-types/issue-25700-1.rs)


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.93.0-nightly (c90bcb957 2025-11-06)
binary: rustc
commit-hash: c90bcb9571b7aab0d8beaa2ce8a998ffaf079d38
commit-date: 2025-11-06
host: x86_64-apple-darwin
release: 1.93.0-nightly
LLVM version: 21.1.3
```

### Error output

Command: `rustc`

```
error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
 --> 08.rs:9:6
  |
9 | impl<T> Clone for S<C>
  |      ^ unconstrained type parameter


thread 'rustc' (6161155) panicked at compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1353:27:
internal error: entered unreachable code
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' (6161155) panicked at compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1353:27:
internal error: entered unreachable code
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: <rustc_borrowck::MirBorrowckCtxt>::suggest_cloning
   4: <rustc_borrowck::MirBorrowckCtxt>::report
   5: <rustc_borrowck::MirBorrowckCtxt>::report_move_errors
   6: rustc_borrowck::borrowck_check_region_constraints
   7: <rustc_borrowck::root_cx::BorrowCheckRootCtxt>::do_mir_borrowck
   8: rustc_borrowck::mir_borrowck
      [... omitted 1 frame ...]
   9: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_interface::passes::run_required_analyses::{closure#1}::{closure#0}>::{closure#0}
  10: rustc_interface::passes::analysis
      [... omitted 1 frame ...]
  11: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
  12: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/Users/jb/Library/CloudStorage/Dropbox/Desk/2503_rustc_ice/20251106 ice 보고용/rustc-ice-2025-11-07T05_57_22-5613.txt` to your bug report

query stack during panic:
#0 [mir_borrowck] borrow-checking `<impl at 08.rs:9:1: 11:23>::clone`
#1 [analysis] running analysis passes on crate `08`
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0207`.
```

</p>
</details>


### Notes
- ICE location: `compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs Line-1353`
https://github.com/rust-lang/rust/blob/c90bcb9571b7aab0d8beaa2ce8a998ffaf079d38/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs#L1346-L1358

- Duplication Check
  - I searched "conflict_errors.rs" ICE ([Github issue page with query](https://github.com/rust-lang/rust/issues?q=is%3Aissue%20state%3Aclosed%20label%3AI-ICE%20%22conflict_errors.rs%22)), and there are no opened issues there.

--- a/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
@@ -1461,15 +1461,19 @@ impl<'infcx, 'tcx> MirBorrowckCtxt<'_, 'infcx, 'tcx> {
             let cause = ObligationCause::misc(expr.span, self.mir_def_id());
             ocx.register_bound(cause, self.infcx.param_env, ty, clone_trait);
             let errors = ocx.evaluate_obligations_error_on_ambiguity();
-            if errors.iter().all(|error| {
-                match error.obligation.predicate.as_clause().and_then(|c| c.as_trait_clause()) {
-                    Some(clause) => match clause.self_ty().skip_binder().kind() {
-                        ty::Adt(def, _) => def.did().is_local() && clause.def_id() == clone_trait,
-                        _ => false,
-                    },
-                    None => false,
-                }
-            }) {
+            if !errors.is_empty()
+                && errors.iter().all(|error| {
+                    match error.obligation.predicate.as_clause().and_then(|c| c.as_trait_clause()) {
+                        Some(clause) => match clause.self_ty().skip_binder().kind() {
+                            ty::Adt(def, _) => {
+                                def.did().is_local() && clause.def_id() == clone_trait
+                            }
+                            _ => false,
+                        },
+                        None => false,
+                    }
+                })
+            {
                 let mut type_spans = vec![];
                 let mut types = FxIndexSet::default();
                 for clause in errors

--- a/tests/ui/moves/unconstrained-impl-param-clone-suggestion.rs
+++ b/tests/ui/moves/unconstrained-impl-param-clone-suggestion.rs
@@ -1,0 +1,20 @@
+// Regression test for https://github.com/rust-lang/rust/issues/148631
+
+struct C;
+
+struct S<T>(T);
+
+trait Tr {}
+
+impl<T> Clone for S<C>
+//~^ ERROR the type parameter `T` is not constrained
+where
+    S<T>: Tr,
+{
+    fn clone(&self) -> Self {
+        *self
+        //~^ ERROR cannot move out of `*self`
+    }
+}
+
+fn main() {}

--- a/tests/ui/moves/unconstrained-impl-param-clone-suggestion.stderr
+++ b/tests/ui/moves/unconstrained-impl-param-clone-suggestion.stderr
@@ -1,0 +1,19 @@
+error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
+  --> $DIR/unconstrained-impl-param-clone-suggestion.rs:9:6
+   |
+LL | impl<T> Clone for S<C>
+   |     -^-
+   |     ||
+   |     |unconstrained type parameter
+   |     help: remove the unused type parameter `T`
+
+error[E0507]: cannot move out of `*self` which is behind a shared reference
+  --> $DIR/unconstrained-impl-param-clone-suggestion.rs:15:9
+   |
+LL |         *self
+   |         ^^^^^ move occurs because `*self` has type `S<C>`, which does not implement the `Copy` trait
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0207, E0507.
+For more information about an error, try `rustc --explain E0207`.

